doc/design/state-machine.md

```diff
@@ -0,0 +1,249 @@
+# HiveD State Machines
+
+This document presents the state machines of HiveD, and explains the life cycles of jobs and resources in HiveD.
+We will describe the state machines of our scheduling unit, Affinity Group, and that of our resource unit, Cell, respectively.
+
+## Affinity Group (AG) State Machine
+
+An affinity group (AG) is a set of gang-scheduled pods. It is the basic scheduling unit in HiveD. The figure below shows the state machine of an affinity group.
+
+Note that the AG state machine has interactions with the cell state machine (elaborated later). In our design, AGs influence each other only via their overlapping cells: for example, an event sent to an AG may trigger an event sent to a cell, and that cell may further trigger another event sent to another AG which is currently associated with the cell.
+Therefore, the state machines of multiple AGs are effectively bridged by the state machines of their overlapping cells.
+
+<p style="text-align: center;">
+  <img src="img/AG-state-machine.png" title="AG" alt="AG" width="70%"/>
+</p>
+
+### States
+
+__`Pending`__: the AG is waiting to be scheduled.
+
+__`Preempting`__: the AG has reserved cells but is waiting for the completion of preemptions of other AGs.
+
+__`Allocated`__: the AG is allocated cells.
```

Comment:
is fully allocated with cells?

Reply:
But what is "partially allocated"? You mean being preempted?

Reply:
Yep, it is one case of only partial his cells are allocated state.
BTW, in below you also used the word fully:
Deleted: the AG is fully deleted and all of its cells are released.